File: solutions/208.implement-trie-prefix-tree.py
#
# @lc app=leetcode id=208 lang=python3
#
# [208] Implement Trie (Prefix Tree)
#

# @lc code=start
class Trie:

    class Node:
        def __init__(self):
            self.is_end_letter = False
            self.children = [None] * 26

    def __init__(self):
        self.root = self.Node()

    def insert(self, word: str) -> None:
        current = self.root
        for c in word:
            index = ord(c) - ord('a')
            if not current.children[index]:
                current.children[index] = self.Node()
            current = current.children[index]
        current.is_end_letter = True

    def search(self, word: str) -> bool:
        last = self._find_last_node(word)
        return last and last.is_end_letter

    def startsWith(self, prefix: str) -> bool:
        return self._find_last_node(prefix)

    def _find_last_node(self, word: str) -> Optional[Node]:
        current = self.root
        for c in word:
            index = ord(c) - ord('a')
            if not current.children[index]:
                return None
            current = current.children[index]
        return current

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

File: solutions/208.implement-trie-prefix-tree.java
/*
 * @lc app=leetcode id=208 lang=java
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start


class Trie {

    Node root;

    class Node {
        boolean isEndLetter = false;
        Node[] children = new Node[26];
        public Node() {}
    }

    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        Node current = root;
        for (char c : word.toCharArray()) {
            if (current.children[c - 'a'] == null) {
                current.children[c - 'a'] = new Node();
            }
            current = current.children[c - 'a'];
        }
        current.isEndLetter = true;
    }
    
    public boolean search(String word) {
        Node node = lastNode(word);
        return node != null && node.isEndLetter;
    }

    public boolean startsWith(String prefix) {
        return lastNode(prefix) != null;
    }

    private Node lastNode(String word) {
        Node current = root;
        for (char c : word.toCharArray()) {
            if (current.children[c - 'a'] == null) {
                return null;
            }
            current = current.children[c - 'a'];
        }
        return current;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
// @lc code=end

